crates/context/config/src/client/near.rs

```diff
@@ -0,0 +1,287 @@
+
+            let _ignored = networks.insert(
+                network_id.clone(),
+                Network {
+                    client,
+                    account_id: network_config.account_id.clone(),
+                    secret_key: network_config.access_key.clone(),
+                },
+            );
+        }
+
+        Self { networks }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("unknown network: {0}")]
+    UnknownNetwork(String),
+    #[error("invalid response from RPC while {operation}")]
+    InvalidResponse { operation: String },
+    #[error("invalid contract ID: {0}")]
+    InvalidContractId(near_primitives::account::id::ParseAccountError),
+    #[error("failed while {operation}: {reason}")]
+    Custom {
+        operation: &'static str,
+        reason: String,
+    },
+}
+
+impl Transport for NearTransport<'_> {
+    type Error = Error;
+
+    async fn send(
+        &self,
+        request: TransportRequest<'_>,
+        payload: Vec<u8>,
+    ) -> Result<Vec<u8>, Self::Error> {
+        let Some(network) = self.networks.get(&request.network_id) else {
+            return Err(Error::UnknownNetwork(request.network_id.into_owned()));
+        };
+
+        let contract_id = request
+            .contract_id
+            .parse()
+            .map_err(Error::InvalidContractId)?;
+
+        match request.operation {
+            Operation::Read { method } => {
+                network
+                    .query(contract_id, method.into_owned(), payload)
+                    .await
+            }
+            Operation::Write { method } => {
+                network
+                    .mutate(contract_id, method.into_owned(), payload)
+                    .await
+            }
+        }
+    }
+}
+
+impl Network {
+    async fn query(
+        &self,
+        contract_id: AccountId,
+        method: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, Error> {
+        let response = self
+            .client
+            .call(methods::query::RpcQueryRequest {
+                block_reference: BlockReference::latest(),
+                request: QueryRequest::CallFunction {
+                    account_id: contract_id,
+                    method_name: method,
+                    args: FunctionArgs::from(args),
+                },
+            })
+            .await
+            .map_err(|err| Error::Custom {
+                operation: "querying contract",
+                reason: err.to_string(),
+            })?;
+
+        match response.kind {
+            QueryResponseKind::CallResult(CallResult { result, logs: _ }) => Ok(result),
+            _ => Err(Error::InvalidResponse {
+                operation: "querying contract".to_owned(),
+            }),
+        }
+    }
+
+    async fn mutate(
+        &self,
+        contract_id: AccountId,
+        method: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, Error> {
```

Comment:
I am not overly fond of this function as it's quite long, and I wonder if it would benefit from being split up? There are a number of distinct stages and areas that could lend themselves well to being individual functions.

Reply:
Extracted out nonce determination